R/to_rmd.R

```diff
@@ -0,0 +1,267 @@
+content_to_rmd <- function(content, output_dir, ..., include_results) {
+  if (include_results || isTRUE(attr(content, "keep"))) {
+    suppressWarnings(hashname <- rlang::hash(content))
+    hashname_file <- paste0(hashname, ".rds")
+    path <- tempfile(fileext = ".rds")
+    suppressWarnings(saveRDS(content, file = path))
+    file.copy(path, file.path(output_dir, hashname_file))
+    sprintf("```{r echo = FALSE}\nreadRDS('%s')\n```", hashname_file)
+  }
+}
+
+#' Convert `ReporterCard`/`doc` content to `rmarkdown`
+#'
+#' This is an S3 generic that is used to generate content in `rmarkdown` format
+#' from various types of blocks in a `ReporterCard` or `doc` object.
+#'
+#' @details
+#' The methods for this S3 generic can be extended by the app developer or even overwritten.
+#' For this a function with the name `to_rmd.<class>` should be defined in the
+#' Global Environment, where `<class>` is the class of the object to be converted.
+#'
+#' For example, to override the default behavior for `code_chunk` class, you can use:
+#'
+#' ```r
+#' to_rmd.code_chunk <- function(block, output_dir, ..., include_results, report_type, eval = TRUE) {
+#'   # custom implementation
+#'   sprintf("### A custom code chunk\n\n```{r}\n%s\n```\n", block)
+#' }
+#' ```
+#'
+#' Alternatively, you can register the S3 method using `registerS3method("to_rmd", "<class>", fun)`
+#'
+#' @param block (`any`) content which can be represented in `rmarkdown` syntax
+#' @param output_dir (`character(1)`) path to the directory where files should be written to.
+#' @return `character(1)` containing a content or `rmarkdown` document
```

Comment:
Would be nice to add some examples. This method might be exported at some point
```suggestion
#' @return `character(1)` containing a content or `rmarkdown` document
#' @examples 
```

Reply:
IIRC CRAN doesn't like `@examples` sections on non-exported functions.

Am I misremembering?

Reply:
Plus, as we're using an internal mechanism for S3 methods, just calling `to_rmd("something")` or even `teal.reporter:::to_rmd("something with 3 double colons")` doesn't work outside the namespace.

``` r
pkgload::load_all("/home/averissimo/work/roche🟦/packages📦/teal.reporter")
#> ℹ Loading teal.reporter
to_rmd("yada")
#> Error in UseMethod("to_rmd"): no applicable method for 'to_rmd' applied to an object of class "character"
teal.reporter:::to_rmd("yada")
#> Error in UseMethod("to_rmd"): no applicable method for 'to_rmd' applied to an object of class "character"
```

<sup>Created on 2025-06-04 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
Good point, let's ignore examples and introduce only when we decide to export these.